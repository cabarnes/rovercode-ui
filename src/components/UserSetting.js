import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { Lock, Person, Settings } from '@material-ui/icons';
import {
  Box,
  Button,
  Grid,
  List,
  ListItem,
  ListItemText,
  TextField,
  Typography,
} from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import { FormattedMessage, injectIntl } from 'react-intl';
import PropTypes from 'prop-types';

class UserSetting extends Component {
  constructor(props) {
    super(props);

    const { user } = props;

    this.state = {
      username: user.username,
      password1: null,
      password2: null,
      saveSuccess: false,
      usernameError: null,
      password1Error: null,
      password2Error: null,
      nonFieldError: null,
    };
  }

  errorMessage = () => {
    const {
      usernameError,
      password1Error,
      password2Error,
      nonFieldError,
    } = this.state;

    const errors = [
      usernameError,
      password1Error,
      password2Error,
      nonFieldError,
    ];

    const activeErrors = errors.filter(Boolean);

    if (activeErrors.length === 0) {
      return (null);
    }

    const messages = [];

    activeErrors.forEach((error) => {
      error.forEach((message) => {
        messages.push(message);
      });
    });

    /* eslint react/no-array-index-key: 0 */
    return (
      <Grid item>
        <Alert variant="outlined" severity="error">
          <List dense>
            {
              messages.map((message, i) => (
                <ListItem key={i}>
                  <ListItemText>
                    {`- ${message}`}
                  </ListItemText>
                </ListItem>
              ))
            }
          </List>
        </Alert>
      </Grid>
    );
  }

  saveUserUsername = (e) => {
    const { editUserUsername } = this.props;
    const { username } = this.state;

    e.preventDefault();

    return editUserUsername(username)
      .then(() => this.setState({ saveSuccess: true }))
      .catch((error) => this.setState({ usernameError: error.response.data.username }));
  }

  saveUserPassword = (e) => {
    const { editUserPassword, intl } = this.props;
    const { password1, password2 } = this.state;

    e.preventDefault();

    const passwordMatch = intl.formatMessage({
      id: 'app.user_setting.match',
      description: 'Error message indicating both passwords must match',
      defaultMessage: 'Passwords must match',
    });

    if (password1 !== password2) {
      this.setState({
        password1Error: [passwordMatch],
        password2Error: [passwordMatch],
      });
      return null;
    }

    return editUserPassword(password1)
      .then(() => this.setState({ saveSuccess: true }))
      .catch((error) => {
        this.setState({
          password1Error: error.response.data.new_password1,
          password2Error: error.response.data.new_password2,
        });
      });
  }

  handleChange = (event) => this.setState({ [event.target.name]: event.target.value })

  render() {
    const { intl, user } = this.props;
    const {
      saveSuccess,
      usernameError,
      password1Error,
      password2Error,
    } = this.state;

    const usernameLabel = intl.formatMessage({
      id: 'app.user_setting.username',
      description: 'Label for username entry',
      defaultMessage: 'Username:',
    });

    const passwordLabel = intl.formatMessage({
      id: 'app.user_setting.password',
      description: 'Label for password entry',
      defaultMessage: 'New Password:',
    });

    const verifyLabel = intl.formatMessage({
      id: 'app.user_setting.verify',
      description: 'Label for password verify entry',
      defaultMessage: 'Verify:',
    });

    return (
      <Grid container direction="column" justify="center" alignItems="center" spacing={2}>
        {
          saveSuccess ? (
            <Redirect to="/accounts/login" />
          ) : (null)
        }
        <Grid item container direction="row" justify="center" alignItems="center" spacing={2}>
          <Grid item>
            <Settings fontSize="large" />
          </Grid>
          <Grid item>
            <Typography variant="h3">
              <FormattedMessage
                id="app.user_setting.header"
                description="Header for the user's settings"
                defaultMessage="My Settings"
              />
            </Typography>
          </Grid>
        </Grid>
        <Grid item>
          <Alert severity="info">
            <FormattedMessage
              id="app.user_setting.info"
              description="Notifies the user that changing settings requires signing back in"
              defaultMessage="Changing any of these settings requires signing back in."
            />
          </Alert>
        </Grid>
        {this.errorMessage()}
        <Grid item>
          <Box p={3} border={1} borderRadius="borderRadius" boxShadow={3} borderColor="grey.500">
            <Grid item container direction="row" justify="center" alignItems="center" spacing={1}>
              <Grid item>
                <Person fontSize="large" />
              </Grid>
              <Grid item>
                <Typography variant="h5">
                  <FormattedMessage
                    id="app.user_setting.username_header"
                    description="Header for changing username"
                    defaultMessage="Change Username"
                  />
                </Typography>
              </Grid>
            </Grid>
            <form key={user.user_id} onSubmit={this.saveUserUsername}>
              <Grid item>
                <TextField
                  variant="outlined"
                  margin="dense"
                  label={usernameLabel}
                  name="username"
                  error={!!usernameError}
                  defaultValue={user.username}
                  onChange={this.handleChange}
                  required
                />
              </Grid>
              <Grid item>
                <Button variant="contained" color="primary" type="submit">
                  <FormattedMessage
                    id="app.user_setting.save_username"
                    description="Button label to save username settings"
                    defaultMessage="Save"
                  />
                </Button>
              </Grid>
            </form>
          </Box>
        </Grid>
        {
          !user.isSocial ? (
            <>
              <Grid item>
                <Box p={3} border={1} borderRadius="borderRadius" boxShadow={3} borderColor="grey.500">
                  <Grid item container direction="row" justify="center" alignItems="center" spacing={1}>
                    <Grid item>
                      <Lock fontSize="large" />
                    </Grid>
                    <Grid item>
                      <Typography variant="h5">
                        <FormattedMessage
                          id="app.user_setting.password_header"
                          description="Header for changing password"
                          defaultMessage="Change Password"
                        />
                      </Typography>
                    </Grid>
                  </Grid>
                  <form key={user.user_id} onSubmit={this.saveUserPassword}>
                    <Grid item container direction="row" spacing={1}>
                      <Grid item>
                        <TextField
                          variant="outlined"
                          margin="dense"
                          label={passwordLabel}
                          name="password1"
                          error={!!password1Error}
                          onChange={this.handleChange}
                          type="password"
                          required
                        />
                      </Grid>
                      <Grid item>
                        <TextField
                          variant="outlined"
                          margin="dense"
                          label={verifyLabel}
                          name="password2"
                          error={!!password2Error}
                          onChange={this.handleChange}
                          type="password"
                          required
                        />
                      </Grid>
                    </Grid>
                    <Grid item>
                      <Button variant="contained" color="primary" type="submit">
                        <FormattedMessage
                          id="app.user_setting.save_password"
                          description="Button label to save password settings"
                          defaultMessage="Save"
                        />
                      </Button>
                    </Grid>
                  </form>
                </Box>
              </Grid>
            </>
          ) : (null)
        }
      </Grid>
    );
  }
}

UserSetting.propTypes = {
  editUserUsername: PropTypes.func.isRequired,
  editUserPassword: PropTypes.func.isRequired,
  user: PropTypes.shape({
    user_id: PropTypes.number.isRequired,
    username: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    isSocial: PropTypes.bool.isRequired,
  }).isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default injectIntl(UserSetting);
