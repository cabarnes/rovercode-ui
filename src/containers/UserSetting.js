import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies, Cookies } from 'react-cookie';
import { editUserPassword, editUserUsername } from '@/actions/user';
import { checkAuthError, authHeader } from '@/actions/auth';
import UserSetting from '@/components/UserSetting';

const mapStateToProps = ({ user }) => ({ user });
const mapDispatchToProps = (dispatch, { cookies }) => ({
  editUserUsername: (username) => dispatch(editUserUsername(username, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
  editUserPassword: (password) => dispatch(editUserPassword(password, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
});

const UserSettingContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(UserSetting);

UserSettingContainer.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default hot(module)(withCookies(UserSettingContainer));
