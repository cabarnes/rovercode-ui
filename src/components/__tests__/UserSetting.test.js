import React from 'react';
import { Redirect } from 'react-router-dom';
import { TextField } from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import UserSetting from '../UserSetting';

let editUserPassword;
let editUserUsername;

describe('The UserSetting component', () => {
  beforeEach(() => {
    editUserPassword = jest.fn(() => Promise.resolve({}));
    editUserUsername = jest.fn(() => Promise.resolve({}));
  });

  test('renders on the page with no errors', () => {
    const user = {
      user_id: 1,
      username: 'testuser',
      email: 'test@example.com',
      isSocial: false,
    };
    const wrapper = shallowWithIntl(
      <UserSetting
        user={user}
        editUserPassword={editUserPassword}
        editUserUsername={editUserUsername}
      />,
    ).dive().dive();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find({ type: 'password' }).length).toBe(2);
  });

  test('hides password edit when social user', () => {
    const user = {
      user_id: 1,
      username: 'testuser',
      email: 'test@example.com',
      isSocial: true,
    };
    const wrapper = shallowWithIntl(
      <UserSetting
        user={user}
        editUserPassword={editUserPassword}
        editUserUsername={editUserUsername}
      />,
    ).dive().dive();
    expect(wrapper.find({ type: 'password' }).exists()).toBe(false);
  });

  test('saves user username', async () => {
    const user = {
      user_id: 1,
      username: 'testuser',
      email: 'test@example.com',
      isSocial: false,
    };
    const wrapper = shallowWithIntl(
      <UserSetting
        user={user}
        editUserPassword={editUserPassword}
        editUserUsername={editUserUsername}
      />,
    ).dive().dive();

    wrapper.find(TextField).first().simulate('change', {
      target: {
        name: 'username',
        value: 'newuser',
      },
    });
    wrapper.update();
    await wrapper.instance().saveUserUsername({ preventDefault: jest.fn() });

    expect(wrapper.state('saveSuccess')).toBe(true);
    expect(wrapper.state('usernameError')).toBeNull();
    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).prop('to')).toBe('/accounts/login');
    expect(editUserPassword).not.toHaveBeenCalled();
    expect(editUserUsername).toHaveBeenCalledWith('newuser');
  });

  test('saves user password', async () => {
    const user = {
      user_id: 1,
      username: 'testuser',
      email: 'test@example.com',
      isSocial: false,
    };
    const wrapper = shallowWithIntl(
      <UserSetting
        user={user}
        editUserPassword={editUserPassword}
        editUserUsername={editUserUsername}
      />,
    ).dive().dive();

    wrapper.find(TextField).at(1).simulate('change', {
      target: {
        name: 'password1',
        value: 'password',
      },
    });
    wrapper.find(TextField).at(2).simulate('change', {
      target: {
        name: 'password2',
        value: 'password',
      },
    });
    wrapper.update();
    await wrapper.instance().saveUserPassword({ preventDefault: jest.fn() });

    expect(wrapper.state('saveSuccess')).toBe(true);
    expect(wrapper.state('password1Error')).toBeNull();
    expect(wrapper.state('password2Error')).toBeNull();
    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).prop('to')).toBe('/accounts/login');
    expect(editUserUsername).not.toHaveBeenCalled();
    expect(editUserPassword).toHaveBeenCalledWith('password');
  });

  test('errors on password mismatch', async () => {
    const user = {
      user_id: 1,
      username: 'testuser',
      email: 'test@example.com',
      isSocial: false,
    };
    const wrapper = shallowWithIntl(
      <UserSetting
        user={user}
        editUserPassword={editUserPassword}
        editUserUsername={editUserUsername}
      />,
    ).dive().dive();

    wrapper.find(TextField).at(1).simulate('change', {
      target: {
        name: 'password1',
        value: '1234',
      },
    });
    wrapper.find(TextField).at(2).simulate('change', {
      target: {
        name: 'password2',
        value: '5678',
      },
    });
    wrapper.update();
    await wrapper.instance().saveUserPassword({ preventDefault: jest.fn() });

    expect(wrapper.state('saveSuccess')).toBe(false);
    expect(wrapper.state('password1Error')).toEqual(['Passwords must match']);
    expect(wrapper.state('password2Error')).toEqual(['Passwords must match']);
    expect(wrapper.find(Alert).exists()).toBe(true);
    expect(wrapper.find(Alert).last().prop('severity')).toBe('error');
    expect(editUserUsername).not.toHaveBeenCalled();
    expect(editUserPassword).not.toHaveBeenCalled();
  });

  test('displays error on username save error', async () => {
    const user = {
      user_id: 1,
      username: 'testuser',
      email: 'test@example.com',
      isSocial: false,
    };
    const error = new Error();
    error.response = {
      status: 400,
      data: {
        username: ['Username taken'],
      },
    };
    editUserUsername = jest.fn(() => Promise.reject(error));
    const wrapper = shallowWithIntl(
      <UserSetting
        user={user}
        editUserPassword={editUserPassword}
        editUserUsername={editUserUsername}
      />,
    ).dive().dive();

    await wrapper.instance().saveUserUsername({ preventDefault: jest.fn() });

    expect(wrapper.state('saveSuccess')).toBe(false);
    expect(wrapper.state('usernameError')).toEqual(['Username taken']);
    expect(wrapper.find(Alert).exists()).toBe(true);
    expect(wrapper.find(Alert).last().prop('severity')).toBe('error');
    expect(wrapper.find(TextField).first().prop('error')).toBe(true);
    expect(wrapper.find(TextField).at(1).prop('error')).toBe(false);
    expect(wrapper.find(TextField).at(2).prop('error')).toBe(false);
  });

  test('displays error on password save error', async () => {
    const user = {
      user_id: 1,
      username: 'testuser',
      email: 'test@example.com',
      isSocial: false,
    };
    const error = new Error();
    error.response = {
      status: 400,
      data: {
        new_password1: ['Too short'],
        new_password2: ['Too short'],
      },
    };
    editUserPassword = jest.fn(() => Promise.reject(error));
    const wrapper = shallowWithIntl(
      <UserSetting
        user={user}
        editUserPassword={editUserPassword}
        editUserUsername={editUserUsername}
      />,
    ).dive().dive();

    await wrapper.instance().saveUserPassword({ preventDefault: jest.fn() });

    expect(wrapper.state('saveSuccess')).toBe(false);
    expect(wrapper.state('password1Error')).toEqual(['Too short']);
    expect(wrapper.state('password2Error')).toEqual(['Too short']);
    expect(wrapper.find(Alert).exists()).toBe(true);
    expect(wrapper.find(Alert).last().prop('severity')).toBe('error');
    expect(wrapper.find(TextField).first().prop('error')).toBe(false);
    expect(wrapper.find(TextField).at(1).prop('error')).toBe(true);
    expect(wrapper.find(TextField).at(2).prop('error')).toBe(true);
  });
});
