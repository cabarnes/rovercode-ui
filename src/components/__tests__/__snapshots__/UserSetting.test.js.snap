// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The UserSetting component renders on the page with no errors 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  container={true}
  direction="column"
  justify="center"
  spacing={2}
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="row"
    item={true}
    justify="center"
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <SettingsIcon
        fontSize="large"
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <WithStyles(ForwardRef(Typography))
        variant="h3"
      >
        <FormattedMessage
          defaultMessage="My Settings"
          id="app.user_setting.header"
          values={Object {}}
        />
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <WithStyles(ForwardRef(Alert))
      severity="info"
    >
      <FormattedMessage
        defaultMessage="Changing any of these settings requires signing back in."
        id="app.user_setting.info"
        values={Object {}}
      />
    </WithStyles(ForwardRef(Alert))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <Styled(MuiBox)
      border={1}
      borderColor="grey.500"
      borderRadius="borderRadius"
      boxShadow={3}
      p={3}
    >
      <WithStyles(ForwardRef(Grid))
        alignItems="center"
        container={true}
        direction="row"
        item={true}
        justify="center"
        spacing={1}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <PersonIcon
            fontSize="large"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Typography))
            variant="h5"
          >
            <FormattedMessage
              defaultMessage="Change Username"
              id="app.user_setting.username_header"
              values={Object {}}
            />
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
      <form
        key="1"
        onSubmit={[Function]}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(TextField))
            defaultValue="testuser"
            error={false}
            label="Username:"
            margin="dense"
            name="username"
            onChange={[Function]}
            required={true}
            variant="outlined"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            type="submit"
            variant="contained"
          >
            <FormattedMessage
              defaultMessage="Save"
              id="app.user_setting.save_username"
              values={Object {}}
            />
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </form>
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <Styled(MuiBox)
      border={1}
      borderColor="grey.500"
      borderRadius="borderRadius"
      boxShadow={3}
      p={3}
    >
      <WithStyles(ForwardRef(Grid))
        alignItems="center"
        container={true}
        direction="row"
        item={true}
        justify="center"
        spacing={1}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <LockIcon
            fontSize="large"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Typography))
            variant="h5"
          >
            <FormattedMessage
              defaultMessage="Change Password"
              id="app.user_setting.password_header"
              values={Object {}}
            />
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
      <form
        key="1"
        onSubmit={[Function]}
      >
        <WithStyles(ForwardRef(Grid))
          container={true}
          direction="row"
          item={true}
          spacing={1}
        >
          <WithStyles(ForwardRef(Grid))
            item={true}
          >
            <WithStyles(ForwardRef(TextField))
              error={false}
              label="New Password:"
              margin="dense"
              name="password1"
              onChange={[Function]}
              required={true}
              type="password"
              variant="outlined"
            />
          </WithStyles(ForwardRef(Grid))>
          <WithStyles(ForwardRef(Grid))
            item={true}
          >
            <WithStyles(ForwardRef(TextField))
              error={false}
              label="Verify:"
              margin="dense"
              name="password2"
              onChange={[Function]}
              required={true}
              type="password"
              variant="outlined"
            />
          </WithStyles(ForwardRef(Grid))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            type="submit"
            variant="contained"
          >
            <FormattedMessage
              defaultMessage="Save"
              id="app.user_setting.save_password"
              values={Object {}}
            />
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </form>
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
